validator/sawtooth_validator/journal/completer.py

```diff
@@ -27,38 +27,51 @@
 from sawtooth_validator.networking.dispatch import Handler
 from sawtooth_validator.networking.dispatch import HandlerResult
 from sawtooth_validator.networking.dispatch import HandlerStatus
+from sawtooth_validator.journal.block_wrapper import NULL_BLOCK_IDENTIFIER
 
 LOGGER = logging.getLogger(__name__)
 
 
 class Completer(object):
-    def __init__(self, block_store):
+    def __init__(self, block_store, gossip):
+        self.gossip = gossip
         self.batch_cache = TimedCache()
         self.block_cache = BlockCache(block_store)
+        # avoid throwing away the genesis block
+        self.block_cache[NULL_BLOCK_IDENTIFIER] = None
         self._on_block_received = None
         self._on_batch_received = None
 
     def _check_block(self, block):
         # currently only accepting finalized blocks
         # in the future if the blocks will be built
-
+        valid = True
         if block.previous_block_id not in self.block_cache:
             LOGGER.debug("Block discarded(Missing predecessor): %s",
                          block.header_signature[:8])
+            LOGGER.debug("Request missing predecessor: %s",
+                         block.previous_block_id)
+            self.gossip.broadcast_block_request(block.previous_block_id)
             return False
+
         if len(block.batches) != len(block.header.batch_ids):
             LOGGER.debug("Block discarded(Missing batches): %s",
                          block.header_signature[:8])
-            return False
+            valid = False
 
-        for i in range(len(block.batches)):
+        for i in range(len(block.header.batch_ids)):
             if block.batches[i].header_signature != block.header.batch_ids[i]:
                 LOGGER.debug("Block discarded(Missing batch): %s",
                              block.header_signature[:8])
-                return False
+                LOGGER.debug("Request missing batch: %s",
+                             block.header.batch_ids[i])
+                self.gossip.broadcast_batch_request(block.header.batch_ids[i],
+                                                    "batch")
+                valid = False
 
-        self.block_cache[block.header_signature] = block
-        return True
+        if valid:
```

Comment:
Even though _check_block() now does things rather than just evaluating conditions and returning true/false, I'd still move this down to the caller, rather than having it as a side effect of _check_block()